File: src/main/java/me/tomassetti/antlrplus/antlrparser/Antlr4ParserFacade.java
package me.tomassetti.antlrplus.antlrparser;

import org.antlr.v4.parse.ANTLRParser;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.TokenStream;

import java.io.*;
import java.nio.charset.StandardCharsets;

public class Antlr4ParserFacade {

    public ANTLRv4Parser.GrammarSpecContext parseString(String code) {
        InputStream inputStream = new ByteArrayInputStream(code.getBytes(StandardCharsets.UTF_8));
        return parseStream(inputStream);
    }

    public ANTLRv4Parser.GrammarSpecContext parseFile(File file) throws FileNotFoundException {
        return parseStream(new FileInputStream(file));
    }

    public ANTLRv4Parser.GrammarSpecContext parseStream(InputStream inputStream) {
        try {
            ANTLRv4Lexer lexer = new ANTLRv4Lexer(new org.antlr.v4.runtime.ANTLRInputStream(inputStream));
            TokenStream tokens = new CommonTokenStream(lexer);
            ANTLRv4Parser parser = new ANTLRv4Parser(tokens);
            return parser.grammarSpec();
        } catch (IOException e) {
            throw new RuntimeException("That is unexpected", e);
        }
    }

}

File: src/main/kotlin/me/tomassetti/antlrplus/antlrparser/Antlr4ParserFacade.kt
package me.tomassetti.antlrplus.antlrparser

import org.antlr.v4.runtime.CommonTokenStream
import java.io.*
import java.nio.charset.StandardCharsets

class Antlr4ParserFacade {

    fun parseString(code: String): ANTLRv4Parser.GrammarSpecContext {
        val inputStream = ByteArrayInputStream(code.toByteArray(StandardCharsets.UTF_8))
        return parseStream(inputStream)
    }

    @Throws(FileNotFoundException::class)
    fun parseFile(file: File): ANTLRv4Parser.GrammarSpecContext {
        return parseStream(FileInputStream(file))
    }

    fun parseStream(inputStream: InputStream): ANTLRv4Parser.GrammarSpecContext {
        try {
            val lexer = ANTLRv4Lexer(org.antlr.v4.runtime.ANTLRInputStream(inputStream))
            val tokens = CommonTokenStream(lexer)
            val parser = ANTLRv4Parser(tokens)
            return parser.grammarSpec()
        } catch (e: IOException) {
            throw RuntimeException("That is unexpected", e)
        }

    }

}
